### ISSUE TITLE
Unexpected quotes passed in when command line contains Chinese and spaces

### ISSUE THREAD

**Issue by fengyichui** [mintty/mintty#1164]
Hi,

Unexpected quotes passed in when command line contains Chinese and spaces.
My command line is:
```bash
> mintty.exe --WSL --exec vim "你 好.txt"   # There is a space between 你 and 好
```
But the filename received in WSL's vim is "你 好.txt" (include leading and trailing quotes), that cause vim can not read the file correctly.

**Comment by mintty** [comment]
You invoke this from a Windows cmd line. Quoting there is weird and following unregular rules. If the quote is passed to the program as part of the argument (as Windows cmd does in presence of non-ASCII characters), cygwin programs get it as such.

**Comment by avih** [comment]
> If the quote is passed to the program as part of the argument (as Windows cmd does in presence of non-ASCII characters), cygwin programs get it as such.

It's true that "cygwin gets it as such" i.e. that this is how cygwin sees the argument, e.g. the following sh script `~/args.sh`:
```sh
#!/bin/sh
printf "[%s]\n" "$@"
```

When invoked like so from windows cmd.exe:
```
C:\>t:\cygwin64\bin\bash ~/args.sh foo bar "你 好.txt"
```

outputs:
```
[foo]
[bar]
["你 好.txt"]
```

Which suggests that indeed cygwin sees the argument with the quotes.

However, that's not how a windows program would see the arguments, e.g. the following C program `args.c`:
```C
    #include <windows.h>
    #include <stdio.h>

    int wmain(int argcw, wchar_t **argvw) {
        for (int i = 1; i < argcw; i++) {
            WriteConsoleW(GetStdHandle(STD_OUTPUT_HANDLE), argvw[i], wcslen(argvw[i]), 0, 0);
            putc('\n', stdout);
        }
        return 0;
    }
```

Compiled with Visual studio from CLI (`cl.exe args.c`), and invoked with the same arguments from the same cmd.exe console
```
c:\args.exe foo bar "你 好.txt"
foo
bar
你 好.txt
```

Which strips the quotes.

So at least as far as windows programs go, quoting chinese chars with spaces at cmd.exe does not seem to introduce a problem.

Assuming cygwin does get used here, then to me it does seem like a cygwin issue rather than an unsolvable issue of cmd.exe, and should probably be reported to cygwin.

However, as this is a WSL invocation, I don't know how cygwin is involved, if at all.

**Comment by avih** [comment]
FWIW, the following program also has the same (correct, stripped quotes) output like args.exe above:
```C
    #include <windows.h>
    #include <shellapi.h>
    #include <stdio.h>

    int main(int c, char **v) {
        int argcw;
        wchar_t **argvw = CommandLineToArgvW(GetCommandLineW(), &argcw);

        for (int i = 1; i < argcw; i++) {
            WriteConsoleW(GetStdHandle(STD_OUTPUT_HANDLE), argvw[i], wcslen(argvw[i]), 0, 0);
            putc('\n', stdout);
        }
        return 0;
    }
```

Compiled like so; `cl.exe args.c shell32.lib`.

This uses `CommandLineToArgvW(GetCommandLineW(), ...)` from `shell32.dll` to parse the line into argv, which as far as I know is technically different than the way in which the arguments get into `argv` of `wmain`.

So that's another way of unicode cli args to be processed correctly in a windows program.

Bottom line, if cygwin is indeed involved, I think it should be able to parse it correctly like windows programs do, and the fact that it seems not to probably means it can be considered a cygwin bug.

**Comment by mintty** [comment]
Thanks, that reminds me that this was indeed discussed on the cygwin mailing list recently.
I think quote handling at the interworking interface from Windows to cygwin has been changed a few times in cygwin history.
I also think there were reasons this could not easily be fixed but I'm not sure.
Anyway, it's not a terminal issue.

**Comment by BrianInglis** [comment]
How `mintty`, *Cygwin*, and *Windows* work are distractions. 
It's likely an @rprichard /*wslbridge* @Biswa96 /*wslbridge2* issue. 
Commands and args are passed thru from front end to the back end over a *pty* or *pipe*, so you have to ensure it passes thru to the front end, and gets handled by the program run on the back end, which has to be a *shell* to handle quoting. 
For example, the *Windows* shortcut command *Target* to start `Xwin Server` is:
```
C:\cygwin64\bin\run.exe --quote /usr/bin/bash.exe -l -c "cd; exec /usr/bin/startxwin"
```
so you may have to jump thru similar hoops using e.g *PowerShell* from *Windows* `cmd` if you want to avoid *Cygwin* `sh`/`bash`.

**Comment by mintty** [comment]
Hi Brian,
it's not a WSL (or bridge) issue; I checked current behaviour with `mintty --exec mined "你 好.txt"`.
There may exist tricky workarounds but they may not work anymore once cygwin establishes its own workaround (again?).

**Comment by BrianInglis** [comment]
If a *Cygwin* process like `mintty` execs a command which is not a shell, then there can be no expectation of shell escape or quote handling: each string is an arg. Such is the case when `mintty`(*child,c*) uses `execv`/`execvp`(3). That is why `run --quote` is used to start `Xwin server`: it quotes args as expected by MS startup, so the shell then gets commands it can use.

**Comment by mintty** [comment]
What I meant is that I called the command above from a Windows cmd, so the inclusion of the quotes into the argument would be done by cmd, I think. A cygwin workaround might be to strip embedding qoutes from arguments when called from outside cygwin (if that can be detected).

**Comment by BrianInglis** [comment]
*Cygwin* expects to run programs from within its root file system, and passes commands and arguments to forked *Cygwin* child processes using its own conventions, so *Cygwin* program startup can detect whether it was forked by *Cygwin*, or run from *Windows*, allowing startup to do some shell-like command line processing on *Windows* command lines, including quote handling, some expansions, and globbing, which it can skip when forked.
